engine/api/src/main/java/io/deephaven/engine/table/Table.java

```diff
@@ -1719,6 +1720,96 @@ Table rollup(Collection<? extends Aggregation> aggregations, boolean includeCons
     @ConcurrentMethod
     Table treeTable(String idColumn, String parentColumn);
 
+    // -----------------------------------------------------------------------------------------------------------------
+    // UpdateBy Operations
+    // -----------------------------------------------------------------------------------------------------------------
+
+    /**
+     * <p>
+     * Create a table with the same {@link #getRowSet() rowSet} as its parent that will perform the specified set of row
+     * based operations to it. As opposed to {@link #update(String...)} these operations are more restricted but are
+     * capable of processing state between rows. This operation will group the table by the specified set of keys if
+     * provided
+     * </p>
+     *
+     * @param operations the operations to apply to the table.
+     * @param byColumns the columns to group by before applying.
+     * @return a table with the same index, with the specified operations applied to each group defined by the
+     *         {@code byColumns}
+     */
+    @ConcurrentMethod
+    Table updateBy(@NotNull final Collection<UpdateByClause> operations, final String... byColumns);
```

Comment:
Collection of polymorphic types  in this context should ideally be `Collections<? extends Type>`, ie `Collection<? extends UpdateByClause>`.